src/hutch_bunny/core/AvailabilityQuerySolver.py

```diff
@@ -0,0 +1,406 @@
+
+    def add_joins_to_person(self):
+        self.condition = self.condition.join(Person, Person.person_id == ConditionOccurrence.person_id)
+        self.drug = self.drug.join(Person, Person.person_id == DrugExposure.person_id)
+        self.meas = self.meas.join(Person, Person.person_id == Measurement.person_id)
+        self.obs = self.obs.join(Person, Person.person_id == Observation.person_id)
+
+    def calc_relative_date(self, left_value_time: str, right_value_time: str) -> datetime:
+
+        time_value_supplied: str
+
+        # have to toggle between left and right, given |1 means less than 1 and
+        # 1| means greater than 1
+        if left_value_time == "":
+            time_value_supplied = right_value_time
+        else:
+            time_value_supplied = left_value_time
+        # converting supplied time (in months) (stored as string) to int, and negating.
+        time_to_use: int = int(time_value_supplied)
+        time_to_use = time_to_use * -1
+
+        # the relative date to search on, is the current date minus
+        # the number of months supplied
+        today_date: datetime = datetime.now()
+        relative_date = today_date + relativedelta(months=time_to_use)
+
+        return relative_date
+
+    def add_person_constraints(self, person_constraints_for_group, current_rule :Rule, concepts):
+        concept_domain: str = concepts.get(current_rule.value)
+
+        if concept_domain == "Gender":
+            if current_rule.operator == "=":
+                person_constraints_for_group.append(Person.gender_concept_id == int(current_rule.value))
+            else:
+                person_constraints_for_group.append(Person.gender_concept_id != int(current_rule.value))
+
+        elif concept_domain == "Race":
+            if current_rule.operator == "=":
+                person_constraints_for_group.append(Person.race_concept_id == int(current_rule.value))
+            else:
+                person_constraints_for_group.append(Person.race_concept_id != int(current_rule.value))
+
+        elif concept_domain == "Ethnicity":
+            if current_rule.operator == "=":
+                person_constraints_for_group.append(Person.ethnicity_concept_id == int(current_rule.value))
+            else:
+                person_constraints_for_group.append(Person.ethnicity_concept_id != int(current_rule.value))
+
+        return person_constraints_for_group
+
+
+    def add_secondary_modifiers(self, current_rule: Rule, condition: select) -> select:
+        # Not sure where, but even when a secondary modifier is not supplied, an array
+        # with a single entry is provided.
+        # todo: need to confirm if this is in the JSON from the API or our implementation
+
+        secondary_modifier_list = list()
+
+
+        for type_index, typeAdd in enumerate(current_rule.secondary_modifier, start=0):
```

Comment:
```suggestion
for typeAdd in current_rule.secondary_modifier:
```
Does not need to be enumerate